cloud_storage: fix request signing with special chars

nvartolomei opened the pull request:
We found an issue where continuation token contains chars which look like url encoded, e.g. plus sign (+). If we send it as-is in the query param but then we url-encode it in the signature the server responds with invalid signature error. That's because it (aws server) unconditionally decodes the query parameters first.

Even if we fixed the request signature it is like that listing won't work as the `+` sign must be sent as `+` and not let the server decode it as a space instead!

So we need to encode query params. This PR does that. And now, we don't need to url encode during signing!

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
cloud_storage: fix request signing with special chars

We found an issue where continuation token contains chars which look
like url encoded, e.g. plus sign (+). If we send it as-is in the query
param but then we url-encode it in the signature the server responds
with invalid signature error. That's because it (aws server)
unconditionally decodes the query parameters first.

Even if we fixed the request signature it is like that listing won't
work as the `+` sign must be sent as `+` and not let the server decode
it as a space instead!

So we need to encode query params. This PR does that. And now, we don't
need to url encode during signing!

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/76312-->
<!--23e95b50-fe88-40bf-9486-cc417fb7e9d2-->
<!--513c31a8-ead1-40e7-8459-1ea2624e391a-->
<!--9ccbbe32-4b5f-4041-ad1c-4f7addfc75af-->
<!--fda84ceb-9281-4e6e-bf3a-48740ae1e22c-->
<!--050d21d9-ae7a-4f89-805b-cfc7d7a89cff-->
<!--c736ded5-acac-4142-b159-b1857f53f508-->
<!--a64a9f0f-8d36-4227-910e-4a059c8aad68-->
<!--42bff45a-511a-48bb-a264-68e8612a3cec-->
<!--2860864d-5372-48cb-9bc7-8e9b0d6eeffa-->
<!--e5fc1b4d-0912-40b5-824c-2e4a93599e5d-->
<!--ffa8626c-0c4a-4528-8fbd-a0d48372b7bf-->
<!--b30fb552-49fe-448d-80df-a83d70f41a88-->
<!--title-->
## Retry command for [Build#76312](https://buildkite.com/redpanda/redpanda/builds/76312)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/e2e_shadow_indexing_test.py::ShadowIndexingInfiniteRetentionTest.test_segments_not_deleted@{"cloud_storage_type":2}
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_many_partitions@{"check_mode":"check_manifest_and_segment_metadata","cloud_storage_type":2}
tests/rptest/tests/cluster_metadata_upload_test.py::ClusterMetadataUploadTest.test_uploads_after_restart@{"cloud_storage_type":2}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingCloudRetentionTest.test_cloud_size_based_retention@{"cloud_storage_type":2}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":false,"topic_remote_write":"true"}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTestWithDisruptions.test_write_with_node_failures@{"cloud_storage_type":2}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_from_cloud@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, AdjacentSegmentMergerReupload == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/archival_test.py::ArchivalTest.test_compaction_topic_config_change@{"cloud_storage_type":2}
tests/rptest/tests/archival_test.py::ArchivalTest.test_reconnect@{"cloud_storage_type":2}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":20}
tests/rptest/tests/adjacent_segment_merging_test.py::AdjacentSegmentMergingTest.test_reupload_of_local_segments@{"acks":1,"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, AdjacentSegmentMergerReupload == True)"}}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingCloudRetentionTest.test_cloud_time_based_retention@{"cloud_storage_type":2}
tests/rptest/tests/remote_label_test.py::RemoteLabelsTest.test_share_bucket_delete_topic@{"cloud_storage_type":2}
tests/rptest/tests/shadow_indexing_admin_api_test.py::SIAdminApiTest.test_bucket_validation@{"cloud_storage_type":2}
tests/rptest/tests/shadow_indexing_firewall_test.py::ShadowIndexingFirewallTest.test_consume_from_blocked_s3@{"cloud_storage_type":2}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_local_time_based_retention_is_overridden@{"cloud_storage_type":2,"local_retention_ms":-1}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":true,"topic_remote_write":null}
tests/rptest/tests/archival_test.py::ArchivalTest.test_connection_drop@{"cloud_storage_type":2}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_reset_spillover@{"cloud_storage_type":2}
tests/rptest/tests/archival_test.py::ArchivalTest.test_retention_archival_coordination@{"acks":-1,"cloud_storage_type":2}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndSpilloverTest.test_spillover@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/adjacent_segment_merging_test.py::AdjacentSegmentMergingToggleCompactionTest.test_reupload_of_local_segments@{"acks":1,"cloud_storage_type":2}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_admin_api_recovery@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_ChunkedRead == True)"}}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingCloudRetentionTest.test_cloud_retention_deleted_segments_count@{"cloud_storage_type":2}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_vcluster_id@{"cloud_storage_type":2}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_gc_entire_manifest@{"cloud_storage_type":2}
tests/rptest/tests/partition_movement_test.py::SIPartitionMovementTest.test_shadow_indexing@{"cloud_storage_type":2,"num_to_upgrade":0,"with_cloud_topics":true}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":true,"topic_remote_write":"false"}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_local_time_based_retention_is_overridden@{"cloud_storage_type":2,"local_retention_ms":3600000}
tests/rptest/tests/e2e_shadow_indexing_test.py::EndToEndShadowIndexingTest.test_write@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True)"}}
tests/rptest/tests/topic_delete_test.py::TopicDeleteCloudStorageTest.drop_lifecycle_marker_test@{"cloud_storage_type":2}
tests/rptest/tests/archival_test.py::ArchivalTest.test_retention_archival_coordination@{"acks":1,"cloud_storage_type":2}
tests/rptest/tests/archival_test.py::ArchivalTest.test_connection_flicker@{"cloud_storage_type":2}
tests/rptest/tests/adjacent_segment_merging_test.py::AdjacentSegmentMergingToggleCompactionTest.test_reupload_of_local_segments@{"acks":-1,"cloud_storage_type":2}
tests/rptest/tests/read_replica_e2e_test.py::TestReadReplicaService.test_writes_forbidden@{"cloud_storage_type":2,"partition_count":10}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_prevent_recovery@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, SegmentRolledByTimeout == True)"}}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingCloudRetentionTest.test_cloud_size_based_retention_application@{"cloud_storage_type":2}
tests/rptest/tests/adjacent_segment_merging_test.py::AdjacentSegmentMergingTest.test_reupload_of_local_segments@{"acks":-1,"cloud_storage_type":2}
tests/rptest/tests/remote_label_test.py::RemoteLabelsTest.test_share_bucket_concurrent_consume@{"cloud_storage_type":2}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.ms"}
tests/rptest/tests/cluster_metadata_upload_test.py::ClusterMetadataUploadTest.test_uploads_after_wipe@{"cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_many_partitions@{"check_mode":"no_check","cloud_storage_type":2}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":false,"topic_remote_write":null}
tests/rptest/tests/retention_policy_test.py::ShadowIndexingLocalRetentionTest.test_shadow_indexing_default_local_retention@{"cloud_storage_type":2,"cluster_remote_write":true,"topic_remote_write":"true"}
tests/rptest/tests/archive_retention_test.py::CloudArchiveRetentionTest.test_delete@{"cloud_storage_type":2,"retention_type":"retention.bytes"}
tests/rptest/tests/shadow_indexing_compacted_topic_test.py::ShadowIndexingCompactedTopicTest.test_upload@{"cloud_storage_type":2}
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.test_many_partitions@{"check_mode":"check_manifest_existence","cloud_storage_type":2}
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[2,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True)"}}
tests/rptest/tests/archival_test.py::ArchivalTest.test_isolate@{"cloud_storage_type":2}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

nvartolomei committed:
http: add url_decode utility

We are permissive on inputs. Invalid encodings are decoded as-is.
Inspired by the AWS SDK.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d854eb10-2d3d-4e07-8650-8c25fe35d16c-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/76324">build#76324</a></summary>


| test_class              | test_method                                  | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------|:---------------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest | test_simple_remount                          | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/76324#019a8356-0adb-44da-ba5c-f7e1fa8680b3 | FLAKY         | 16/21    | upstream reliability is '81.70289855072464'. current run reliability is '76.19047619047619'. drift is 5.51242 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
| SegmentMsTest           | test_segment_rolling_with_retention_consumer | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/76324#019a8356-0d00-4ac8-b82c-deb699ce9c5f | FLAKY         | 19/21    | upstream reliability is '92.47467438494935'. current run reliability is '90.47619047619048'. drift is 1.99848 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<!--/body-->

nvartolomei commented:
/cdt
rp_version=latest
tests/rptest/tests/nodes_decommissioning_test.py

vbotbuildovich commented:
Invalid input. When rp_repo=ci, the rp_version should be: build or pr

[Workflow run logs][1].

[1]: https://github.com/redpanda-data/redpanda/actions/runs/19377296066

nvartolomei commented:
/cdt
rp_version=pr
tests/rptest/tests/nodes_decommissioning_test.py

nvartolomei commented:
/cdt
provider=azure
region=centralus
cdt_instance_type=Standard_L8as_v3
rp_version=pr
tests/rptest/tests/nodes_decommissioning_test.py

andrwng commented on src/v/cloud_roles/signature.cc:
```diff
@@ -467,7 +467,10 @@ result<ss::sstring> signature_abs::get_canonicalized_resource(
     for (const auto& [pname, pvalue] : query_params) {
         // TODO(vlad): List values are not supported.
-        result += ssx::sformat("\n{}:{}", pname, pvalue);
+        // https://learn.microsoft.com/en-us/rest/api/storageservices/authorize-with-shared-key#shared-key-format-for-2009-09-19-and-later
```
👍

vbotbuildovich commented:
/backport v25.3.x
